src/Prompt.php

```diff
@@ -154,37 +156,30 @@ public function newLinesWritten(): int
      */
     protected function capturePreviousNewLines(): void
     {
-        $this->newLinesWritten = method_exists(static::output(), 'newLinesWritten')
-            ? static::output()->newLinesWritten()
-            : 1;
+        $this->newLinesWritten = static::writer()->newLinesWritten();
     }
 
     /**
      * Set the output instance.
      */
     public static function setOutput(OutputInterface $output): void
     {
-        self::$output = $output;
+        if ($output instanceof ConsoleOutputInterface && stream_isatty(STDERR) && ! stream_isatty(STDOUT)) {
+            $output = $output->getErrorOutput();
+        }
+        self::$writer = new PromptWriter($output);
```

Comment:
Laravel sends an instance of `\Illuminate\Console\OutputStyle`, which doesn't implement `ConsoleOutputInterface`, meaning we won't use STDERR with Laravel even when STDOUT is redirected elsewhere.

Laravel's `OutputStyle` class is an extension of Symfony's `SymfonyStyle` and `OutputStyle` classes, which are a wrapper around an underlying `OutputInterface`. It doesn't expose the `getErrorOutput` method but instead exposes a `getErrorStyle` method. Unfortunately, this returns a new instance of `SymfonyStyle` rather than Laravel's `OutputStyle` class (due to their use of `self`).

Even if we could get it to return a new instance of Laravel's `OutputStyle`, the newline tracking would break because Laravel will still be using its original `OutputStyle` instance rather than the new instance we'd be using in Prompts.

It's a complicated problem because we need to keep track of the newlines between Prompts's and Laravel's output. Both Laravel and Prompts need to know how many trailing newlines were in the previous output, regardless of where that previous output came from, so that they each know how many newlines to emit before any new output. If Laravel isn't aware of the trailing newlines that Prompts has written, it won't be able to output the correct amount of leading newlines before any new output, and vice versa. It gets more complicated with Symfony because they have a separate instance for STDERR. If Laravel and Prompts use different instances, they won't be aware of each other's trailing newlines.

The only solutions I can think of are:

1. Add logic to Laravel to make it use an STDERR output instance when appropriate (and share that instance with Prompts). This could have unintended consequences in user commands that expect to output on STDOUT regardless of any redirection, so it's probably a breaking change. Alternatively,
2. Add a method to Laravel's `OutputStyle` class (or create a new wrapper class) that allows us to write to STDERR, while still tracking trailing newlines regardless of the stream. It would potentially need to be smart enough to track newlines separately when streams are redirected to different places because, in that scenario, the trailing newlines from the previous output of another stream shouldn't be considered. Prompts could then choose to output to STDERR and both Laravel and Prompts would still know how many leading newlines they needed to output.

Either way it's a big change covering two code bases. I still see the value in doing this, but I'm concerned about the effort and risk involved.

Reply:
Thanks for the detailed explanation. I can't think of any other solutions, so you're welcome to close this if you want.
I learned a lot about console output, and it was nice to meet you guys 😄

Reply:
Thanks, @mortenscheel! Nice to meet you as well :slightly_smiling_face: 

Happy to reconsider this at some point if we can figure out a good solution. Having one blank line between output looks great, but it certainly adds a lot of challenges :sweat_smile: